### PR TITLE
p2p: always advertise self, to enable mutual address discovery

### LINKED ISSUE #7593: p2p: mutual address discovery doesn't happen (on v0.35.x)
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): d0f01e9ace0e9ebf4a254d8743b3d6527e4584c2

**ABCI app** (name for built-in, URL for self-written if it's publicly available): vocdoni.io

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:
i set up a testnet, with 1 seed and 2 fullnodes (validators) that connect to that seed.
each node stays connected to the seed only, and don't connect to each other directly

**What you expected to happen**:
i expected that both fullnodes would connect to the seed, the seed shares the addresses with both and the fullnodes connect to each other directly

i dug through the code, and concluded that when the seed gets an incoming connection from `node0`, doesn't have any way to know the "diallable" address of `node0`. there's even a comment in the code about it in `peermanager.go` Accepted():

```go
// NOTE: We can't take an address here, since e.g. TCP uses a different port
// number for outbound traffic than inbound traffic, so the peer's endpoint
// wouldn't necessarily be an appropriate address to dial.
//
// FIXME: When we accept a connection from a peer, we should register that
// peer's address in the peer store so that we can dial it later. In order to do
// that, we'll need to get the remote address after all, but as noted above that
// can't be the remote endpoint since that will usually have the wrong port
// number.
```

so the peerStore from the seed is not populated with `node0` and `node1` addresses, and thus replies with an empty pex_response.

### PR DESCRIPTION
**Pull request by altergui**
Fixes #7593

### CODE REVIEW

**Comment by altergui** [comment]
(i also prepared [hotfix/pex-advertise-self_v0.35.x](https://github.com/vocdoni/tendermint/tree/hotfix/pex-advertise-self_v0.35.x) to apply on top of `v0.35.x`)

File: internal/p2p/peermanager.go [hunk 1]
```diff
@@ -271,7 +272,7 @@ type PeerManager struct {
 }
 
 // NewPeerManager creates a new peer manager.
-func NewPeerManager(selfID types.NodeID, peerDB dbm.DB, options PeerManagerOptions) (*PeerManager, error) {
+func NewPeerManager(selfID types.NodeID, externalAddr string, peerDB dbm.DB, options PeerManagerOptions) (*PeerManager, error) {
```
Comment (tychoish) [inline internal/p2p/peermanager.go:275]: Seems like we should add this option to the `PeerManagerOptions` struct rather than passing in another argument.

**Review by tychoish — CHANGES_REQUESTED** [review]
is it possible to write a test for this

File: internal/p2p/peermanager.go (hunk 1 above)
Comment (altergui) [reply internal/p2p/peermanager.go:275]: perfect, will do. i actually considered this originally and took a look at PeerManagerOptions but for whatever reason had the feeling it didn't quite fit, possibly just because i was too tired :laughing:

**Comment by altergui** [comment]
i think i can write a test for this

thanks a lot for the review!!

File: internal/p2p/peermanager.go (hunk 1 above)
Comment (altergui) [reply internal/p2p/peermanager.go:275]: solved in e38f0be5ba227347f89b203d25a77e959cc3e0be

**Comment by altergui** [comment]
just pushed v2, much cleaner.

still need to write a test for this

**Comment by altergui** [comment]
added a simple `TestPeerManager_Advertise_Self()`

i did notice that the behaviour of advertise changes subtly in an edge case, if SelfAddress is defined:
when asked for 0 peers (as in `peerManager.Advertise(aID, 0)`) it will return, well, 1 peer (itself). `TestPeerManager_Advertise()` still PASSes ok because SelfAddress is not defined.

**Review by tychoish — APPROVED** [review]
I think this is fine. I have a nagging suspicion that seed nodes might want to avoid promoting themselves (and some operators may eventually ask to be able to redact themselves,) but I think its fine to test the waters with this.

**Review by creachadair — APPROVED** [review]

**Comment by tychoish** [comment]
I think this needs to be rebased, but I added the automerge flag.

**Comment by tychoish** [comment]
It looks like we're out of date again: I think mergify (or us?) needs access to be able to push to your fork to be able to rebase/update before merging.

**Comment by altergui** [comment]
oh, i just pushed a rebase, and now read your comment
